i2c/ssd1306_i2c/ssd1306_i2c.c

```diff
@@ -0,0 +1,348 @@
+    // byte 1 is x = 1, y=0->7 etc
+
+    // This code could be optimised, but is like this for clarity. The compiler
+    // should do a half decent job optimising it anyway.
+
+    const int BytesPerRow = 128; // 128 pixels, 1bpp, but each row is 8 pixel high, so (128 / 8) * 8
+
+    int byte_idx = (y / 8) * BytesPerRow   +   x;
+    uint8_t byte = Framebuffer[byte_idx];
+
+    if (on)
+        byte |=  1 << (y % 8);
+    else
+        byte &= ~(1 << (y % 8));
+
+    Framebuffer[byte_idx] = byte;
+}
+
+// Basic Bresenhams.
+static void DrawLine(int x0, int y0, int x1, int y1, bool on) {
+
+    int dx =  abs(x1-x0);
+    int sx = x0<x1 ? 1 : -1;
+    int dy = -abs(y1-y0);
+    int sy = y0<y1 ? 1 : -1;
+    int err = dx+dy;
+    int e2;
+
+    while (true) {
+        SetPixel(x0, y0, on);
+
+        if (x0 == x1 && y0 == y1)
+            break;
+        e2 = 2*err;
+
+        if (e2 >= dy)
+        {
+            err += dy;
+            x0 += sx;
+        }
+        if (e2 <= dx)
+        {
+            err += dx;
+            y0 += sy;
+        }
+    }
+}
+
+static uint8_t reverse(uint8_t b) {
+   b = (b & 0xF0) >> 4 | (b & 0x0F) << 4;
+   b = (b & 0xCC) >> 2 | (b & 0x33) << 2;
+   b = (b & 0xAA) >> 1 | (b & 0x55) << 1;
+   return b;
+}
+
+
+static inline int GetFontIndex(uint8_t ch) {
+    if (ch >= 'A' && ch <='Z')
+        return  ch - 'A' + 1;
+    else if (ch >= '0' && ch <='9')
+        return  ch - '0' + 27;
+    else
+        return  0; // Not got that char so space.
+}
+
+static uint8_t reversed[sizeof(font)] = {0};
+
+static void FillReversedCache() {
+    // calculate and cache a reversed version of fhe font, because I defined it upside down...doh!
+    for (int i=0;i<sizeof(font);i++)
+        reversed[i] = reverse(font[i]);
+}
+
+static void WriteChar(uint x, uint y, uint8_t ch) {
+    if (reversed[0] == 0)
+        FillReversedCache();
+
+    if (x > SSD1306_LCDWIDTH - 8 || y > SSD1306_LCDHEIGHT - 8)
+        return;
+
+    // For the moment, only write on Y row boundaries (every 8 vertical pixels)
+    y = y/8;
+
+    ch = toupper(ch);
+    int idx = GetFontIndex(ch);
+    int fb_idx = y * 128 + x;
+
+    for (int i=0;i<8;i++) {
+        Framebuffer[fb_idx++] = reversed[idx * 8 + i];
+    }
+}
+
+static uint16_t ExpandByte(uint8_t b) {
+    uint16_t w = 0;
+    for (int i=7;i>=0;i--) {
+        uint16_t t = (b & (1 << i));
+        w |= (t << i);
+        w |= (t << (i + 1));
+    }
+    return w;
+}
+
+static void WriteBigChar(uint x, uint y, uint8_t ch) {
+    if (reversed[0] == 0)
+        FillReversedCache();
+
+    if (x > SSD1306_LCDWIDTH - 16 || y > SSD1306_LCDHEIGHT - 16)
+        return;
+
+    // For the moment, only write on Y row boundaries (every 8 vertical pixels)
+    y = y/8;
+
+    ch = toupper(ch);
+    int idx = GetFontIndex(ch);
+    int fb_idx = y * 128 + x;
+
+    for (int i=0;i<8;i++) {
+        uint16_t w = ExpandByte(reversed[idx * 8 + i]);
```

Comment:
RP2040 has lots of memory - maybe it's worth creating a BigFontCache rather than calling ExpandByte over and over again? :man_shrugging:

Reply:
Certainly a possibility. I'll have think as it's quite an expensive function, but tbh, the display refresh massively overwhelmed the rendering time. Its only 512  bytes of cache though.....